### PR TITLE
[IMP] website: improve dynamic snippet

### PR DESCRIPTION
**Pull request by detrouxdev**
**WILL BE BACKPORTED IN 16.0**

The goal of this pull request is to improve the dynamic snippet so that templates have more flexibility.

### CODE REVIEW

File: addons/website/static/src/snippets/s_dynamic_snippet/000.js [hunk 1]
```diff
@@ -149,12 +151,17 @@ const DynamicSnippet = publicWidget.Widget.extend({
...
     _getQWebRenderOptions: function () {
+        const dataset = this.$target[0].dataset;
+        const numberOfRecords = parseInt(dataset.numberOfRecords);
+        let numberOfElements;
+        if (config.device.isMobile) {
+            numberOfElements = parseInt(dataset.numberOfElementsSmallDevices) || DEFAULT_NUMBER_OF_ELEMENTS_SM
+        } else {
+            numberOfElements = parseInt(dataset.numberOfElements) || DEFAULT_NUMBER_OF_ELEMENTS;
+        }
+        let chunkSize = numberOfRecords < numberOfElements ? numberOfRecords : numberOfElements;
```
Comment (bso-odoo) [inline addons/website/static/src/snippets/s_dynamic_snippet/000.js:162]: `const` ?

File: addons/website/static/src/snippets/s_dynamic_snippet/options.js [hunk 2]
```diff
@@ -47,15 +47,6 @@ const dynamicSnippetOptions = options.Class.extend({
...
             this._templateUpdated(widgetValue, params.activeValue);
         }
-        if (params.attributeName === 'numberOfRecords' && previewMode === false) {
-            const dataSet = this.$target.get(0).dataset;
-            if (dataSet.numberOfElements > dataSet.numberOfRecords) {
-                dataSet.numberOfElements = dataSet.numberOfRecords;
-            }
-            if (dataSet.numberOfElementsSmallDevices > dataSet.numberOfRecords) {
-                dataSet.numberOfElementsSmallDevices = dataSet.numberOfRecords;
-            }
```
Comment (bso-odoo) [inline addons/website/static/src/snippets/s_dynamic_snippet/options.js:57]: The problem of the missing `parseInt` in number comparisons should be done in a stable version (15.0 ?)

File: addons/website/static/src/snippets/s_dynamic_snippet/000.js [hunk 3]
```diff
@@ -149,12 +151,17 @@ const DynamicSnippet = publicWidget.Widget.extend({
...
      */
     _getQWebRenderOptions: function () {
+        const dataset = this.$target[0].dataset;
+        const numberOfRecords = parseInt(dataset.numberOfRecords);
+        let numberOfElements;
+        if (config.device.isMobile) {
+            numberOfElements = parseInt(dataset.numberOfElementsSmallDevices) || DEFAULT_NUMBER_OF_ELEMENTS_SM
+        } else {
+            numberOfElements = parseInt(dataset.numberOfElements) || DEFAULT_NUMBER_OF_ELEMENTS;
+        }
```
Comment (bso-odoo) [inline addons/website/static/src/snippets/s_dynamic_snippet/000.js:161]: Could be a `const` by using a ternary operator. It is up to you, just a matter of taste.

File: addons/website/static/src/snippets/s_dynamic_snippet_carousel/000.xml [hunk 4]
```diff
@@ -1,36 +1,59 @@
...
-                    <div t-attf-class="carousel-item #{slideIndex_first ? 'active' : ''}">
-                        <div class="row">
-                            <t t-foreach="itemIndexGenerator" t-as="itemIndex">
-                                <t t-if="(slideIndex * chunkSize + itemIndex) &lt; data.length">
-                                    <div t-attf-class="#{colClass}">
-                                        <t t-out="data[slideIndex * chunkSize + itemIndex]"/>
-                                    </div>
-                                </t>
+                    <t t-set="slideOffset" t-value="slideIndex * slideSize"/>
+                    <div t-attf-class="carousel-item #{extraClasses} #{slideIndex_first ? 'active' : ''}">
```
Comment (bso-odoo) [inline addons/website/static/src/snippets/s_dynamic_snippet_carousel/000.xml:24]: If find no other reference to `slideIndex_first`. Maybe `slideIndex == 0` would be an equivalent condition to what was intended ?

File: addons/website/static/src/snippets/s_dynamic_snippet_carousel/000.xml [hunk 5]
```diff
@@ -1,36 +1,59 @@
...
+        </a>
+        <a t-attf-href="##{uniqueId}" t-attf-class="o_w_carousel_control carousel-control-next" data-slide="next" role="button" aria-label="Next" title="Next">
+            <span class="fa fa-chevron-circle-right fa-2x"/>
+            <span class="sr-only">Next</span>
+        </a>
+    </t>
     <t t-name="website.s_dynamic_snippet.carousel">
         <div t-att-id="uniqueId" class="carousel slide" t-att-data-interval="interval">
             <!-- Content -->
+            <t t-set="slideSize" t-value="chunkSize * rowPerSlide" />
```
Comment (bso-odoo) [inline addons/website/static/src/snippets/s_dynamic_snippet_carousel/000.xml:16]: What about renaming `chunkSize` to `rowSize` ? Or if not renamed outside this template, assigning it to a `rowSize` variable ?

File: addons/website/static/src/snippets/s_dynamic_snippet_carousel/options.js [hunk 6]
```diff
@@ -18,7 +18,30 @@ const dynamicSnippetCarouselOptions = s_dynamic_snippet_options.extend({
...
+        } else {
+            delete this.$target[0].dataset.rowPerSlide;
+        }
+        
+        if (template.arrowPosition) {
+            this.$target[0].dataset.arrowPosition = template.arrowPosition;
+        } else {
+            delete this.$target[0].dataset.arrowPosition;
+        }
+
```
Comment (bso-odoo) [inline addons/website/static/src/snippets/s_dynamic_snippet_carousel/options.js:43]: Empty line.

File: addons/website/views/snippets/s_dynamic_snippet_carousel.xml [hunk 7]
```diff
@@ -9,7 +9,6 @@
         <t t-call="website.s_dynamic_snippet_options_template">
             <t t-set="snippet_name" t-value="snippet_name"/>
             <t t-set="snippet_selector" t-value="snippet_selector"/>
-            <t t-set="grouping_message">Items per slide</t>
```
Comment (bso-odoo) [inline addons/website/views/snippets/s_dynamic_snippet_carousel.xml:12]: Note that this line still survived in `s_dynamic_snippet.xml` and `s_events.xml`. I guess it should also be removed there.

File: addons/website/static/src/snippets/s_dynamic_snippet/000.js [hunk 8]
```diff
@@ -149,12 +151,17 @@ const DynamicSnippet = publicWidget.Widget.extend({
      * @private
      */
     _getQWebRenderOptions: function () {
+        const dataset = this.$target[0].dataset;
+        const numberOfRecords = parseInt(dataset.numberOfRecords);
+        let numberOfElements;
+        if (config.device.isMobile) {
+            numberOfElements = parseInt(dataset.numberOfElementsSmallDevices) || DEFAULT_NUMBER_OF_ELEMENTS_SM
```
Comment (bso-odoo) [inline addons/website/static/src/snippets/s_dynamic_snippet/000.js:158]: Missing semi-colon at the end of the line.

File: addons/website/static/src/snippets/s_dynamic_snippet/000.js [hunk 9]
```diff
@@ -177,6 +180,7 @@ const DynamicSnippet = publicWidget.Widget.extend({
             this.renderedContent = '';
         }
         this._renderContent();
+        this.trigger_up('widgets_start_request', { $target: this.$el.children(), options: {parent: this} });
```
Comment (bso-odoo) [inline addons/website/static/src/snippets/s_dynamic_snippet/000.js:183]: No space after `{` nor before `}`.

File: addons/website_sale/static/src/js/website_sale_utils.js [hunk 10]
```diff
@@ -53,12 +53,14 @@ function animateClone($cart, $elem, offsetTop, offsetLeft) {
...
                     left: $imgtodrag.offset().left
                 })
+                .removeClass('w-100 h-100 h-auto w-auto')
                 .addClass('o_website_sale_animate')
                 .appendTo(document.body)
                 .css({
                     // Keep the same size on cloned img.
                     width: $imgtodrag.width(),
                     height: $imgtodrag.height(),
+                    objectFit: $imgtodrag.css("object-fit")
```
Comment (bso-odoo) [inline addons/website_sale/static/src/js/website_sale_utils.js:63]: Should this be backported ?

**Review by bso-odoo — CHANGES_REQUESTED** [review]
Here are a few suggestions.
Also typos in:
- 4th commit message: "let ... returning" => "let ... return"
- last commit message: "multiples rows" => "multiple rows".

File: addons/website/static/src/snippets/s_dynamic_snippet/options.js [hunk 11]
```diff
@@ -224,10 +215,8 @@ const dynamicSnippetOptions = options.Class.extend({
             if (!this.dynamicFilterTemplates[selectedTemplateId]) {
                 this.$target.get(0).dataset['templateKey'] = dynamicFilterTemplates[0].key;
                 this.isOptionDefault['templateKey'] = true;
-                setTimeout(() => {
-                    this._templateUpdated(dynamicFilterTemplates[0].key, selectedTemplateId);
-                    this._refreshPublicWidgets();
-                });
+                this._templateUpdated(dynamicFilterTemplates[0].key, selectedTemplateId);
+                this._refreshPublicWidgets();
```
Comment (bso-odoo) [inline addons/website/static/src/snippets/s_dynamic_snippet/options.js:203]: `[2]` being a fix already applied on 15.0, shouldn't removing the `setTimeout` also be done on 15.0 ?

**Comment by detrouxdev** [comment]
@bso-odoo Thanks for your review! I agree that the fix should be back-ported to 15.0. I made a PR here #82222. To be noted that it isn't the exact same fix and that I will probably implement the code made here into the commit that gives more flexibility to the template.

File: addons/website/static/src/snippets/s_dynamic_snippet_carousel/000.xml (hunk 4 above)
Comment (detrouxdev) [reply addons/website/static/src/snippets/s_dynamic_snippet_carousel/000.xml:24]: I'm not sure I understand this. As per [Odoo Docs](https://www.odoo.com/documentation/15.0/developer/reference/frontend/qweb.html#loops):
> $as_first
whether the current item is the first of the iteration (equivalent to $as_index == 0)

Maybe this doesn't work in QWeb JS?

File: addons/website_sale/static/src/js/website_sale_utils.js (hunk 10 above)
Comment (detrouxdev) [reply addons/website_sale/static/src/js/website_sale_utils.js:63]: I think this is related to the new templates being added. I don't seem to find these specific classes on IMGs in 15.0. Maybe I'm wrong?

cc @Brieuc-brd

File: addons/website/static/src/snippets/s_dynamic_snippet_carousel/000.xml (hunk 5 above)
Comment (detrouxdev) [reply addons/website/static/src/snippets/s_dynamic_snippet_carousel/000.xml:16]: I'm always careful about renaming variables that were already there. It might break things that I'm not thinking of or customisation done by others. 
However maybe for the sake of readability here, I should do a `t-set="rowSize" t-value="chunkSize"` and replace chunkSize by rowSize in the template. Was that what you're suggesting?

File: addons/website/static/src/snippets/s_dynamic_snippet_carousel/000.xml (hunk 4 above)
Comment (bso-odoo) [reply addons/website/static/src/snippets/s_dynamic_snippet_carousel/000.xml:24]: Oops, I had forgotten about this, ignore my comment.
QWeb.js seems to support it:
```js
context[val_first] = j == 0;
```

File: addons/website/static/src/snippets/s_dynamic_snippet_carousel/000.xml (hunk 5 above)
Comment (bso-odoo) [reply addons/website/static/src/snippets/s_dynamic_snippet_carousel/000.xml:16]: Yes, this is what I had in mind.

File: addons/website_sale/static/src/js/website_sale_utils.js (hunk 10 above)
Comment (Brieuc-brd) [reply addons/website_sale/static/src/js/website_sale_utils.js:66]: @detrouxdev 
It allows to insure to have a perfect clone of the image for the animation where the image is dragged to the cart. The clone is created when we click on the button and it is removed at the end of the animation.

[https://www.awesomescreenshot.com/video/6771423?key=4d9f534679c0b3e69e6d3618de508303](url)

**Comment by detrouxdev** [comment]
I cherry-picked 2 commits that are fixes from #82222 for a smoother testing

File: addons/web_editor/static/src/js/wysiwyg/wysiwyg.js [hunk 12]
```diff
@@ -76,6 +76,7 @@ const Wysiwyg = Widget.extend({
         this.$editable.data('oe-id', options.recordInfo.res_id);
         document.addEventListener('mousedown', this._onDocumentMousedown, true);
         this.$editable.on('blur', this._onBlur);
+        $('.o_not_editable').attr('contentEditable', false);
```
Comment (dmo-odoo) [inline addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:79]: Don't do this. See discussion at https://github.com/odoo/odoo/pull/82222#discussion_r818819590 with the solution being currently developed at https://github.com/odoo/odoo/pull/87347

Comment (qsm-odoo) [reply addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:79]: We know, that's from here that the discussion started :wink: This task is currently waiting for that fix to be merged indeed :wink:

Comment (dmo-odoo) [reply addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:79]: Oh alright, sorry then. It's getting quite hard to follow the maze of different PRs on this subject... ^^'

File: addons/website/static/src/snippets/s_dynamic_snippet_carousel/000.scss [hunk 13]
```diff
@@ -18,4 +18,9 @@
     text-overflow: ellipsis;
     overflow: hidden;
   }
+  .o_w_dynamic_snippet_arrow_bottom {
+    > .o_w_carousel_control {
+      position: relative;
+    }
+  }
```
Comment (bvr-odoo) [inline addons/website/static/src/snippets/s_dynamic_snippet_carousel/000.scss:25]:
```suggestion
  .s_dynamic_snippet_arrow_bottom {
    > .carousel-control-prev, > .carousel-control-next {
      position: relative;
    }
  }
 ```

File: addons/website_sale/data/product_snippet_template_data.xml [hunk 14]
```diff
@@ -234,13 +247,12 @@
                                 </div>
                                 <div class="card-text text-muted" t-field="record.description_sale"/>
                                 <div class="mt-4">
-                                    <a class="btn btn-outline-secondary me-1 mt-1" t-att-href="record.website_url">
+                                    <button t-if="record._website_show_quick_add()" type="button" role="button" class="btn btn-primary js_add_cart mt-1" title="Add to Cart">
+                                        Add to Cart
+                                    </button>
+                                    <a class="btn btn-link mr-1 mt-1" t-att-href="record.website_url">
```
Comment (bvr-odoo) [inline addons/website_sale/data/product_snippet_template_data.xml:253]: `mr-1` => `me-1` (Boostrap V5) maybe see if there are no other cases in the pr

File: addons/website_sale/data/product_snippet_template_data.xml [hunk 15]
```diff
@@ -250,42 +262,129 @@
...
         <template id="dynamic_filter_template_product_product_horizontal_card" name="Horizontal Card">
-            <t t-foreach="records" t-as="data" data-number-of-elements="3" data-number-of-elements-sm="1">
+            <t t-foreach="records" t-as="data"
+                data-number-of-elements="3"
+                data-number-of-elements-sm="1"
+                data-row-per-slide="2"
+                data-arrow-position="bottom"
+                data-extra-classes="o_carousel_multiple_rows"
+                data-thumb="/website_sale/static/src/img/snippets_options/product_horizontal_card.svg"
+                >
```
Comment (bvr-odoo) [inline addons/website_sale/data/product_snippet_template_data.xml:272]:
```suggestion
data-thumb="/website_sale/static/src/img/snippets_options/product_horizontal_card.svg">
```

File: addons/website_sale/static/src/snippets/s_dynamic_snippet_products/000.scss [hunk 16]
```diff
@@ -35,8 +35,52 @@
...
+        @include border-top-start-radius($border-radius-lg !important);
+    }
+    & > .row:first-child > div:last-child .o_carousel_product_card {
+        @include border-top-end-radius($border-radius-lg !important);
+    }
+    & > .row:last-child > div:first-child .o_carousel_product_card {
+        @include border-bottom-start-radius($border-radius-lg !important);
+    }
+    & > .row:last-child > div:last-child .o_carousel_product_card {
+        @include border-bottom-end-radius($border-radius-lg !important);
```
Comment (bvr-odoo) [inline addons/website_sale/static/src/snippets/s_dynamic_snippet_products/000.scss:84]: the `&` are no needed here

Comment (bvr-odoo) [reply addons/website_sale/static/src/snippets/s_dynamic_snippet_products/000.scss:84]: several selectors can be grouped here... (e.g. `> .row:first-child >`, `> .row`)

Comment (detrouxdev) [reply addons/website_sale/static/src/snippets/s_dynamic_snippet_products/000.scss:84]:
```scss
.o_card_group {
    > .row {
        padding-left: $grid-gutter-width / 2;
        padding-right: $grid-gutter-width / 2;
        > div {
            padding: 0 !important;
            &:first-child > .o_carousel_product_card{
                border-left: $card-border-width solid $card-border-color !important;
            }
        }
        &:first-child {
            .o_carousel_product_card {
                border-top: $card-border-width solid $card-border-color !important;
            }
            > div {
                &:first-child .o_carousel_product_card {
                    @include border-top-start-radius($border-radius-lg !important);
                }
                &:last-child .o_carousel_product_card {
                    @include border-top-end-radius($border-radius-lg !important);
                }
            }
        }
        &:last-child {
            > div {
                &:first-child .o_carousel_product_card {
                    @include border-bottom-start-radius($border-radius-lg !important);
                }
                &:last-child .o_carousel_product_card {
                    @include border-bottom-end-radius($border-radius-lg !important);
                }
            }
        }
    }
}
```

Would this be better? What do you think?

Comment (bvr-odoo) [reply addons/website_sale/static/src/snippets/s_dynamic_snippet_products/000.scss:84]: > ```css-scss
> .o_card_group {
>     > .row {
>         padding-left: $grid-gutter-width / 2;
>         padding-right: $grid-gutter-width / 2;
>         > div {
>             padding: 0 !important;
>             &:first-child > .o_carousel_product_card{
>                 border-left: $card-border-width solid $card-border-color !important;
>             }
>         }
>         &:first-child {
>             .o_carousel_product_card {
>                 border-top: $card-border-width solid $card-border-color !important;
>             }
>             > div {
>                 &:first-child .o_carousel_product_card {
>                     @include border-top-start-radius($border-radius-lg !important);
>                 }
>                 &:last-child .o_carousel_product_card {
>                     @include border-top-end-radius($border-radius-lg !important);
>                 }
>             }
>         }
>         &:last-child {
>             > div {
>                 &:first-child .o_carousel_product_card {
>                     @include border-bottom-start-radius($border-radius-lg !important);
>                 }
>                 &:last-child .o_carousel_product_card {
>                     @include border-bottom-end-radius($border-radius-lg !important);
>                 }
>             }
>         }
>     }
> }
> ```
> 
> Would this be better? What do you think?

yes it's better like that ;)

File: addons/website/static/src/snippets/s_dynamic_snippet/000.js [hunk 17]
```diff
@@ -13,6 +13,9 @@ const DynamicSnippet = publicWidget.Widget.extend({
     read_events: {
         'click [data-url]': '_onCallToAction',
     },
+    custom_events: {
+        's_dynamic_snippet_rerender': '_onRerender',
+    },
```
Comment (qsm-odoo) [inline addons/website/static/src/snippets/s_dynamic_snippet/000.js:18]: Is that needed? Can't you just restart the public widget entirely?

**Review by qsm-odoo — CHANGES_REQUESTED** [review]
First batch of review (checked the first 2 commits), will continue reviewing now but at least you can already start adapting / answering if needed :+1:

File: addons/website/controllers/main.py [hunk 18]
```diff
@@ -401,6 +401,7 @@ def get_dynamic_snippet_templates(self, filter_name=False):
             t['rowPerSlide'] = attribs.get('data-row-per-slide')
             t['arrowPosition'] = attribs.get('data-arrow-position')
             t['extraClasses'] = attribs.get('data-extra-classes')
+            t['thumb'] = attribs.get('data-thumb')
```
Comment (qsm-odoo) [inline addons/website/controllers/main.py:404]: Review all commits one by one: when you edit a commit message, you have the list of apps you touched. I can't see a commit message saying "website_sale" and have the first change in "website" :wink:

File: addons/website_sale/data/product_snippet_template_data.xml [hunk 19]
```diff
@@ -3,18 +3,18 @@
     <data>
         <!-- Templates for Dynamic Snippet -->
         <template id="dynamic_filter_template_product_product_add_to_cart" name="Classic Card">
-            <t t-foreach="records" t-as="data">
+            <t t-foreach="records" t-as="data" data-thumb="/website_sale/static/src/img/snippets_options/product_add_to_cart.svg">
```
Comment (qsm-odoo) [inline addons/website_sale/data/product_snippet_template_data.xml:6]: How can this work? A `data-XXX` on a `<t>` instruction... that probably does nothing, right? :thinking:

File: addons/website_sale/data/product_snippet_template_data.xml [hunk 20]
```diff
@@ -23,78 +23,78 @@
                                 <t t-set="rating_count" t-value="record.rating_count"/>
                             </t>
                         </div>
-                    </div>
-                    <div class="o_carousel_product_card_footer card-footer d-flex align-items-center">
-                        <div>
-                            <t t-call="website_sale.price_dynamic_filter_template_product_product"/>
+                        <div class="w-100 d-flex flex-wrap flex-md-column flex-lg-row align-items-center align-self-end justify-content-between mt-3">
```
Comment (qsm-odoo) [inline addons/website_sale/data/product_snippet_template_data.xml:26]: Just making sure: has this been retested since BS5?

File: addons/website_sale/data/product_snippet_template_data.xml [hunk 21]
```diff
@@ -23,78 +23,78 @@
...
                 </div>
             </t>
         </template>
 
-        <template id="dynamic_filter_template_product_product_view_detail" name="Detailed Product">
-            <t t-foreach="records" t-as="data">
-                <t t-set="record" t-value="data['_record']"/>
-                <div class="o_carousel_product_card card h-100 w-100" t-att-data-url="record.website_url" t-att-data-add2cart-rerender="data.get('_add2cart_rerender')">
+        <template id="dynamic_filter_template_product_product_view_detail" name="Classic Card - Detailed">
+            <t t-foreach="records" t-as="data" data-number-of-elements="3" data-thumb="/website_sale/static/src/img/snippets_options/product_view_detail.svg">
```
Comment (qsm-odoo) [inline addons/website_sale/data/product_snippet_template_data.xml:42]: Same (everywhere else too)

File: addons/website/static/src/snippets/s_dynamic_snippet/options.js [hunk 22]
```diff
@@ -4,9 +4,18 @@ odoo.define('website.s_dynamic_snippet_options', function (require) {
...
+    /**
+     * @typedef {Object} Template - definition of a dynamic snippet template
+     * @property {string} key - key of the template
+     * @property {string} numOfEl - number of elements on desktop
+     * @property {string} numOfElSm - number of elements on mobile
+     * @property {string} numOfElFetch - number of elements to fetch
+     * @property {string} rowPerSlide - number of rows per slide
+     * @property {string} arrowPosition - position of the arrows
+     * @property {string} extraClasses - classes to be added to the <section>
+     */
```
Comment (qsm-odoo) [inline addons/website/static/src/snippets/s_dynamic_snippet/options.js:16]: What does this documents? :thinking:

File: addons/website/static/src/snippets/s_dynamic_snippet/options.js [hunk 23]
```diff
@@ -4,9 +4,18 @@ odoo.define('website.s_dynamic_snippet_options', function (require) {
 const options = require('web_editor.snippets.options');
 
 const dynamicSnippetOptions = options.Class.extend({
+    /**
```
Comment (qsm-odoo) [inline addons/website/static/src/snippets/s_dynamic_snippet/options.js:7]: > Prior to this commit, the Dynamic Snippet options would modify the DOM
at places were it's not supposed to. 

Is this a fix ? Stable ? Or is it an imp and it's ok without this in stable ? I think we talked about it but can't really remember, remind me :)

File: addons/website/static/src/snippets/s_dynamic_snippet/options.js [hunk 24]
```diff
@@ -17,16 +26,27 @@ const dynamicSnippetOptions = options.Class.extend({
...
+    },
     /**
      *
      * @override
      */
-    onBuilt: function () {
-        this._setOptionsDefaultValues();
+    async onBuilt() {
+        await this._setOptionsDefaultValues();
+        return this._refreshPublicWidgets();
```
Comment (qsm-odoo) [inline addons/website/static/src/snippets/s_dynamic_snippet/options.js:49]: Is that actually needed? Maybe add comments?

File: addons/website/static/src/snippets/s_dynamic_snippet/options.js [hunk 25]
```diff
@@ -71,6 +91,10 @@ const dynamicSnippetOptions = options.Class.extend({
     // Private
     //--------------------------------------------------------------------------
 
+    /**
+     * @returns {Template}
+     * @private
```
Comment (qsm-odoo) [inline addons/website/static/src/snippets/s_dynamic_snippet/options.js:96]: `@`private first

File: addons/website/static/src/snippets/s_dynamic_snippet_carousel/000.xml [hunk 26]
```diff
@@ -11,7 +11,7 @@
         </a>
     </t>
     <t t-name="website.s_dynamic_snippet.carousel">
-        <div t-att-id="uniqueId" class="carousel slide" t-att-data-bs-interval="interval">
+        <div t-att-id="uniqueId" class="carousel slide" t-att-data-bs-interval="interval" data-bs-ride="carousel">
```
Comment (qsm-odoo) [inline addons/website/static/src/snippets/s_dynamic_snippet_carousel/000.xml:14]: Is that a "FIX"? "would only start sliding after the first interaction.", you mean that only when you hovered it once, right? If yes, that was not a bug but a choice (between the 2 bootstrap options) and this should be marked as an IMP (change in behavior)?

**Review by qsm-odoo — CHANGES_REQUESTED** [review]

File: addons/website/static/src/snippets/s_dynamic_snippet_carousel/000.xml [hunk 27]
```diff
@@ -1,36 +1,60 @@
...
+    <t t-name="website.s_dynamic_snippet.carousel.arrows">
+        <a t-attf-href="##{uniqueId}" class="carousel-control-prev" data-bs-slide="prev" role="button" aria-label="Previous" title="Previous">
+            <span class="fa fa-chevron-circle-left fa-2x"/>
+            <span class="visually-hidden">Previous</span>
+        </a>
+        <a t-attf-href="##{uniqueId}" class="carousel-control-next" data-bs-slide="next" role="button" aria-label="Next" title="Next">
+            <span class="fa fa-chevron-circle-right fa-2x"/>
+            <span class="visually-hidden">Next</span>
+        </a>
+    </t>
```
[the remark this answers is not in the document]
Comment (detrouxdev) [reply addons/website/static/src/snippets/s_dynamic_snippet_carousel/000.xml:12]: I've tried the following multiple times:
- Create a DB in master with website_sale
- Drop a product snippet, select a template customize it
- Install my branch and run odoo with -u website,website_sale

The new templates are available, each card will have a new design, but the customization will remain (number of cards displayed on the screen)
The user can select a new template and it will apply the default options. They can also select the same template to have the default options re-applied.

Only thing that might not work is custom templates.
TL;DR default templates should work fine, unsure about custom templates, though I think they should work fine as well.

File: addons/website/static/src/snippets/s_dynamic_snippet/000.js (hunk 17 above)
[the remark this answers is not in the document]
Comment (detrouxdev) [reply addons/website/static/src/snippets/s_dynamic_snippet/000.js]: It is indeed a indeed a fix. I targeted master with this one as I couldn't find a solution that could also work in stable. I also saw this fix as more of a "Nice to have" than a fix since it targets a debug-only snippet.

I can try again to find a way to make it work in stable if you want?

Comment (detrouxdev) [reply addons/website/static/src/snippets/s_dynamic_snippet/000.js]: I made changed
```js
this.trigger_up('s_dynamic_snippet_rerender')
```
to
```js
this.trigger_up('widgets_start_request', {target: this.$el.closest('.s_dynamic')})
```
does it seem alright to you?

File: addons/website_sale/data/product_snippet_template_data.xml (hunk 19 above)
Comment (detrouxdev) [reply addons/website_sale/data/product_snippet_template_data.xml:6]: Getting infos from the template is indeed special, see https://github.com/odoo/odoo/blob/a00d53d9fff027f1f2d50687a005f176ea45adcd/addons/website/controllers/main.py#L389

I only built on top of it to add extra data-attributes, such as a thumbnails for the templates inside the option.

File: addons/website_sale/data/product_snippet_template_data.xml (hunk 20 above)
Comment (detrouxdev) [reply addons/website_sale/data/product_snippet_template_data.xml:26]: I did adapt it to BS5. Though I wasn't the one who made those templates, maybe it's worth if @Brieuc-brd can have a final look?

File: addons/website/static/src/snippets/s_dynamic_snippet/options.js (hunk 22 above)
Comment (detrouxdev) [reply addons/website/static/src/snippets/s_dynamic_snippet/options.js:16]: This documents the template infos that we get from the `/website/snippet/filter_templates` route. I then reuse it to define the `this.dynamicFilterTemplates` property.

I honestly hesitated to leave it, but thought that it might be worth it since it's not always clear what the route sends back to you.

Do you want me to remove it? Or add comments to make it clearer what it represents?

File: addons/website/static/src/snippets/s_dynamic_snippet_carousel/000.xml (hunk 27 above)
[the remark this answers is not in the document]
Comment (qsm-odoo) [reply addons/website/static/src/snippets/s_dynamic_snippet_carousel/000.xml:12]: :+1:

File: addons/website/static/src/snippets/s_dynamic_snippet/000.js (hunk 17 above)
[the remark this answers is not in the document]
Comment (qsm-odoo) [reply addons/website/static/src/snippets/s_dynamic_snippet/000.js]: Let's merge this here and we'll see in stable later. Can't the problematic button just be hidden in JS in the case it does not work maybe, something like that is always ok in stable (if fixed in master).

Comment (qsm-odoo) [reply addons/website/static/src/snippets/s_dynamic_snippet/000.js]: If it works, seems better indeed :+1:

File: addons/website_sale/data/product_snippet_template_data.xml (hunk 19 above)
Comment (qsm-odoo) [reply addons/website_sale/data/product_snippet_template_data.xml:6]: Grmbl, ok :) I guess there will be more room for improvements in 16.X :)

File: addons/website_sale/data/product_snippet_template_data.xml (hunk 20 above)
Comment (qsm-odoo) [reply addons/website_sale/data/product_snippet_template_data.xml:26]: I would not mind having his opinion but we are a bit too rushed to delay this further (I am already not a fan a merging it now but since some people really wants it...). Let's not delay the merge (I'll merge tonight or tomorrow max), I was merely making sure you had a look at potential BS5 problems :+1:

File: addons/website/static/src/snippets/s_dynamic_snippet/options.js (hunk 23 above)
Comment (qsm-odoo) [reply addons/website/static/src/snippets/s_dynamic_snippet/options.js:7]: Maybe mention that in the commit message (the PR that was for 15.0, the fact it was decided to do it master and why, etc :wink:)

**Review by qsm-odoo — APPROVED** [review]
@robodoo rebase-ff r+
